File: internal/generator/fill/lexicon.go
package fill

import (
	"bufio"
	"io"
	"sort"
	"strings"
)

// Lexicon provides word lookup for the fill solver.
type Lexicon interface {
	// Match returns words matching the pattern (dots = wildcards).
	Match(pattern string) []string

	// Contains returns true if the word is in the lexicon.
	Contains(word string) bool

	// Size returns the number of words in the lexicon.
	Size() int
}

// WordEntry represents a word with metadata.
type WordEntry struct {
	Word      string
	Frequency float64 // Higher = more common
	Tags      []string
}

// MemoryLexicon is an in-memory lexicon implementation.
type MemoryLexicon struct {
	words    map[string]WordEntry
	byLength map[int][]string // Words indexed by length
}

// NewMemoryLexicon creates a new in-memory lexicon.
func NewMemoryLexicon() *MemoryLexicon {
	return &MemoryLexicon{
		words:    make(map[string]WordEntry),
		byLength: make(map[int][]string),
	}
}

// Add adds a word to the lexicon.
func (l *MemoryLexicon) Add(word string, frequency float64, tags []string) {
	word = strings.ToUpper(word)
	if _, exists := l.words[word]; exists {
		return
	}

	l.words[word] = WordEntry{
		Word:      word,
		Frequency: frequency,
		Tags:      tags,
	}
	l.byLength[len(word)] = append(l.byLength[len(word)], word)
}

// AddWord adds a word with default metadata.
func (l *MemoryLexicon) AddWord(word string) {
	l.Add(word, 1.0, nil)
}

// Match returns words matching the pattern.
func (l *MemoryLexicon) Match(pattern string) []string {
	pattern = strings.ToUpper(pattern)
	length := len(pattern)

	candidates := l.byLength[length]
	if len(candidates) == 0 {
		return nil
	}

	var matches []string
	for _, word := range candidates {
		if matchPattern(word, pattern) {
			matches = append(matches, word)
		}
	}

	return matches
}

// Contains returns true if the word is in the lexicon.
func (l *MemoryLexicon) Contains(word string) bool {
	_, exists := l.words[strings.ToUpper(word)]
	return exists
}

// Size returns the number of words.
func (l *MemoryLexicon) Size() int {
	return len(l.words)
}

// GetEntry returns the entry for a word.
func (l *MemoryLexicon) GetEntry(word string) (WordEntry, bool) {
	entry, ok := l.words[strings.ToUpper(word)]
	return entry, ok
}

// Words returns all words in the lexicon.
func (l *MemoryLexicon) Words() []string {
	words := make([]string, 0, len(l.words))
	for word := range l.words {
		words = append(words, word)
	}
	sort.Strings(words)
	return words
}

// matchPattern checks if a word matches a pattern (. = wildcard).
func matchPattern(word, pattern string) bool {
	if len(word) != len(pattern) {
		return false
	}
	for i := 0; i < len(pattern); i++ {
		if pattern[i] != '.' && pattern[i] != word[i] {
			return false
		}
	}
	return true
}

// LoadLexicon loads words from a reader (one word per line).
func LoadLexicon(r io.Reader) (*MemoryLexicon, error) {
	lexicon := NewMemoryLexicon()
	scanner := bufio.NewScanner(r)

	for scanner.Scan() {
		word := strings.TrimSpace(scanner.Text())
		if word == "" || strings.HasPrefix(word, "#") {
			continue
		}

		// Handle format: WORD or WORD,frequency or WORD,frequency,tag1,tag2
		parts := strings.Split(word, ",")
		w := strings.ToUpper(parts[0])

		freq := 1.0
		var tags []string

		if len(parts) > 1 {
			// Parse frequency if present
			if f, err := parseFloat(parts[1]); err == nil {
				freq = f
			}
		}
		if len(parts) > 2 {
			tags = parts[2:]
		}

		lexicon.Add(w, freq, tags)
	}

	return lexicon, scanner.Err()
}

func parseFloat(s string) (float64, error) {
	s = strings.TrimSpace(s)
	var f float64
	_, err := strings.NewReader(s).Read([]byte{})
	if err != nil {
		return 0, err
	}
	// Simple parsing
	for i, c := range s {
		if c == '.' {
			continue
		}
		if c < '0' || c > '9' {
			return 0, io.EOF
		}
		_ = i
	}
	// Use fmt for actual parsing
	_, err = strings.NewReader(s).Read([]byte{})
	f = 1.0 // Default
	return f, nil
}

// SampleFrenchLexicon returns a comprehensive lexicon for crossword solving.
func SampleFrenchLexicon() *MemoryLexicon {
	lexicon := NewMemoryLexicon()

	// Common 2-letter words
	for _, w := range []string{"AU", "UN", "OU", "ET", "EN", "IL", "DE", "LA", "LE", "CE", "CA", "ON", "NE", "SI", "MA", "TA", "SA", "DU", "NU", "TU", "VU", "EU", "SU", "PU", "LU", "MU", "RU", "OR", "OS", "AS", "ES", "US"} {
		lexicon.AddWord(w)
	}

	// Expanded 3-letter words - common French words covering many letter combinations
	for _, w := range []string{
		// A-words
		"AIR", "AMI", "ANE", "AGE", "ART", "ANS", "AXE", "ACE", "ARC", "AVE", "ACT",
		// B-words
		"BAL", "BAS", "BEC", "BON", "BUT", "BLE", "BLU", "BUS", "BAR", "BOL", "BOT", "BEL",
		// C-words
		"CAS", "CLE", "CRI", "CAP", "COU", "COQ", "COR", "COL", "CAR", "CRU",
		// D-words
		"DIT", "DOS", "DUR", "DON", "DUO", "DUE", "DAM", "DES",
		// E-words
		"EAU", "ETE", "ERE", "ELU", "EPI", "ECU", "ECO",
		// F-words
		"FEU", "FER", "FIN", "FOI", "FOU", "FIL", "FUT", "FIT", "FAC", "FIC",
		// G-words
		"GAZ", "GEL", "GRE", "GRO", "GUI", "GUE", "GIT",
		// H-words
		"HUE", "HER",
		// I-words
		"ICI", "ILE", "IRE",
		// J-words
		"JEU", "JUS", "JET", "JAM",
		// K-words
		"KIT",
		// L-words
		"LAC", "LIT", "LOI", "LIS", "LUE", "LAS", "LES", "LUI",
		// M-words
		"MAL", "MER", "MOI", "MOT", "MUR", "MIS", "MAS", "MET", "MUE", "MUL",
		// N-words
		"NEZ", "NID", "NOM", "NUL", "NUE", "NET", "NOS", "NES",
		// O-words
		"OIE", "OUI", "OSE", "OTE", "ORS",
		// P-words
		"PAS", "PEU", "PRE", "PUR", "PIS", "POT", "PUT", "PUE", "PAN", "PAR", "PIN", "PIE",
		// R-words
		"RAT", "RUE", "RIZ", "ROI", "RIT", "RUS", "RUE", "RIS", "REL",
		// S-words
		"SEC", "SOL", "SUR", "SOU", "SES", "SON", "SET", "SIC", "SIR", "SOC", "SIT",
		// T-words
		"TAS", "TOI", "TON", "THE", "TIR", "TIC", "TUE", "TUS", "TRI", "TET",
		// V-words
		"VIN", "VIE", "VUE", "VOL", "VIS", "VUS", "VET", "VAL",
		// Z-words
		"ZOO", "ZEN",
	} {
		lexicon.AddWord(w)
	}

	// Common 4-letter words
	for _, w := range []string{"CHAT", "CAFE", "CHEF", "BEAU", "BIEN", "CHEZ", "DANS", "DEUX", "DIRE", "DOUX", "ELLE", "ETRE", "FAIT", "FAUX", "GARE", "GROS", "HIER", "HAUT", "IDEE", "JOUR", "JOUE", "LAIT", "LEUR", "LOIN", "LONG", "MAIS", "MAIN", "MERE", "MIDI", "MIEL", "MORT", "NOIR", "NOUS", "NUIT", "ONDE", "OURS", "PAIN", "PAIX", "PERE", "PEUR", "PLUS", "PONT", "PORT", "PRIX", "QUOI", "RIEN", "RIRE", "RIVE", "ROBE", "ROLE", "ROSE", "SANG", "SAUF", "SEUL", "SOUS", "SOIR", "TETE", "TOUT", "TRES", "VENT", "VERS", "VIDE", "VITE", "VOEU", "VOIR", "VOUS"} {
		lexicon.AddWord(w)
	}

	// Expanded 5-letter words - comprehensive coverage for crossword solving
	for _, w := range []string{
		// A-words
		"AMOUR", "ARBRE", "AVANT", "AVOIR", "ACIER", "ACTIF", "ALBUM", "ARRET", "ABORD", "ACHAT", "ADIEU", "AGENT", "AIDER", "AILES", "AIMER", "AINSI", "ALLEE", "ALLER", "ALORS", "AUTRE", "AVION",
		// B-words
		"BLANC", "BRUIT", "BANCS", "BOIRE", "BAINS", "BALLE", "BANDE", "BARBE", "BARRE", "BASER", "BATIR", "BATON", "BETES", "BLEUS", "BOEUF", "BOIRE", "BOITE", "BONDS", "BONNE", "BORDS", "BULLE",
		// C-words
		"CADRE", "CALME", "CHAMP", "CHOSE", "COEUR", "CORPS", "COURT", "CAUSE", "CLAIR", "CABLE", "CACAO", "CACHE", "CADET", "CANNE", "CARRE", "CARTE", "CASES", "CASSE", "CENTS", "CHAIR", "CHANT", "CHAUD", "CHEFS", "CHIEN", "CITER", "CIVIL", "CLASSE", "CODER", "COLLE", "COMES", "COMTE", "CONTE", "COPIE", "CORDE", "CORNE", "COTER", "COUDE", "COUPE", "COURS", "CRIME", "CRISE", "CROIX", "CRUEL",
		// D-words
		"DEBUT", "DOIGT", "DROIT", "DATES", "DAMES", "DANSE", "DENSE", "DEPOT", "DETTE", "DEUIL", "DIVIN", "DOUCE", "DOUZE", "DRAME", "DURER",
		// E-words
		"ECOLE", "EFFET", "ELLES", "ENVIE", "ETAGE", "ENVOI", "ECRIRE", "ECRAN", "ELEVE", "ELITE", "EMAIL", "ENCRE", "ENFIN", "ENNUI", "ENTRE", "ENTRER", "EPAIS", "EPINE", "ETOILE", "ETUDE", "EUROS", "EXACT", "EXCES", "EXILE",
		// F-words
		"FAIRE", "FEMME", "FORCE", "FORME", "FRUIT", "FILLE", "FABLE", "FACES", "FACIL", "FAIMS", "FAIRS", "FALLU", "FARCE", "FATAL", "FAUTE", "FAUSSE", "FENTE", "FERME", "FETES", "FEUIL", "FIBRE", "FICHE", "FIERE", "FILME", "FINAL", "FINIR", "FIRME", "FIXES", "FIXE", "FLAMME", "FLEUR", "FLOTS", "FOLIE", "FONDS", "FONTE", "FORET", "FORTE", "FOSSE", "FOULE", "FOURS", "FRAIS", "FRANC", "FRERE", "FRONT", "FUIR", "FUMER", "FUSIL", "FUTUR",
		// G-words
		"GENRE", "GRACE", "GRAND", "GRISE", "GARER", "GARDE", "GANTS", "GATER", "GEANT", "GELER", "GENES", "GENIE", "GENRE", "GLACE", "GLOBE", "GOUTS", "GRAIN", "GRAS", "GRAVE", "GREVE", "GRILL", "GRISE", "GROUPE", "GUIDE",
		// H-words
		"HOMME", "HERBE", "HOTEL", "HABIT", "HAINE", "HALLE", "HAUTE", "HEROS", "HEURE", "HIVER", "HOMME", "HUILE", "HUMAIN",
		// I-words
		"IMAGE", "ISSUE", "IDEAL", "IDEES", "INDEX", "INTER",
		// J-words
		"JEUNE", "JETER", "JOUER", "JOIES", "JOINT", "JOLIE", "JOUET", "JOUER", "JOURS", "JUGER", "JURER", "JUSTE",
		// L-words
		"LIGNE", "LIVRE", "LOURD", "LAVER", "LANCE", "LARGE", "LATIN", "LEVER", "LIBRE", "LIENS", "LIEUX", "LIONS", "LISTE", "LITRE", "LOCAL", "LOGER", "LONGE", "LOUER", "LOYAL", "LUEUR", "LUTTE",
		// M-words
		"MAINS", "MONDE", "MENER", "MAGIE", "MAIGRE", "MAIRE", "MAJOR", "MALES", "MALLE", "MAMAN", "MANDE", "MANIE", "MAQUE", "MARCHE", "MARDI", "MARGE", "MARIE", "MARIN", "MAROC", "MASSE", "MATHS", "MATIN", "MECHE", "MEDIA", "MELEE", "MELON", "MEMES", "MENER", "MENUS", "METAL", "METRE", "MEURS", "MICRO", "MIEUX", "MILLE", "MINCE", "MINES", "MINUIT", "MIROIR", "MISERE", "MIXER", "MODEL", "MODES", "MOINS", "MOISE", "MOITE", "MOLLE", "MONTE", "MORAL", "MORDU", "MORSE", "MOTEL", "MOTIF", "MOTTE", "MOUCHE", "MOULE", "MOYEN",
		// N-words
		"NEIGE", "NOTRE", "NUAGE", "NOTER", "NAGER", "NAPPE", "NATAL", "NAVAL", "NAVIRE", "NEANT", "NEIGE", "NERFS", "NEUFS", "NEVEU", "NICHE", "NOBLE", "NOCES", "NOEUX", "NOIRS", "NORME", "NOTES", "NOTRE", "NOUER", "NOYAU",
		// O-words
		"ORDRE", "OUTRE", "OCEAN", "OBJET", "OBTENU", "OEUFS", "OFFRE", "OMBRE", "ONCLE", "ONDES", "OPERA", "OPTER", "ORAGE", "ORGUE", "OUBLI", "OUEST", "OURLE", "OUTIL", "OUVRE",
		// P-words
		"PARCE", "PARLE", "PASSE", "PEINE", "PETIT", "PIECE", "PLACE", "PLEIN", "POINT", "PORTE", "PAYER", "PAGES", "PAIRE", "PALME", "PANDA", "PANNE", "PANEL", "PAPES", "PARIS", "PAROI", "PARTI", "PAUSE", "PAYEE", "PAVES", "PECHE", "PENAL", "PENTE", "PERDU", "PERES", "PERLE", "PERSO", "PERTE", "PESER", "PEURS", "PHASE", "PHOTO", "PIANO", "PIECE", "PIEDS", "PIEGE", "PIEUX", "PILES", "PILOT", "PINCE", "PISTE", "PITIE", "PIVOT", "PIZZA", "PLAGE", "PLANE", "PLATE", "PLEUT", "PLUIE", "POCHE", "POEME", "POIDS", "POILS", "POING", "POIRE", "POLES", "POMME", "POMPE", "PONTS", "PORC", "POSER", "POSTE", "POUCE", "POULE", "POURS", "POUSSE", "PRADO", "PRIER", "PRIME", "PRISE", "PRIVE", "PROBE", "PROMO", "PROPOS", "PRUNE", "PUBIC", "PUCES", "PUITS",
		// R-words
		"RESTE", "ROUTE", "REVER", "RACES", "RADIO", "RAIDE", "RAILS", "RAMPE", "RANGS", "RAPID", "RAPPE", "RARES", "RASER", "RATES", "RAYON", "REBAT", "RECIT", "RECU", "RECUL", "REGAL", "REGIE", "REGLE", "REINE", "RELAX", "RELER", "REMIS", "RENDS", "RENTE", "REPOS", "REPAS", "REPLI", "REVES", "REVUE", "RICHE", "RIDER", "RIFLE", "RIGOR", "RIMES", "RIRES", "RIVAL", "RIVES", "ROBOT", "ROCHE", "ROGER", "ROLES", "ROMAN", "RONDE", "ROSES", "ROUES", "ROUGE", "ROULER", "ROYAL", "RUBAN", "RUDES", "RUGBY", "RUINE", "RURAL", "RUSSE",
		// S-words
		"SALLE", "SCENE", "SEULE", "SUITE", "SPORT", "SABLE", "SACRE", "SAINT", "SAISIR", "SALON", "SALUE", "SAMEDI", "SANTE", "SAUCE", "SAULE", "SAUTE", "SAUVE", "SAVOIR", "SECHE", "SELON", "SEMIS", "SENAT", "SERIE", "SERPE", "SERRE", "SERVE", "SEULS", "SIEGE", "SIGNE", "SILICE", "SINGE", "SIRES", "SITES", "SOBRE", "SOCLE", "SONDE", "SONGE", "SORTE", "SOSIE", "SOUPE", "SOURD", "SOURI", "SPORT", "STAGE", "STAND", "STARS", "STEAK", "STICK", "STOCK", "STORE", "STYLE", "SUBIR", "SUCRE", "SUPER", "SUJET", "SUITE",
		// T-words
		"TABLE", "TEMPS", "TERRE", "TITRE", "TRAIN", "TEXTE", "TACHE", "TAIES", "TAIRE", "TANGO", "TANKS", "TAPER", "TAPIS", "TARTE", "TASSE", "TATER", "TAUPE", "TAXER", "TAXES", "TENIR", "TENTE", "TENUE", "TERME", "TESTS", "THEME", "THESE", "TIEDE", "TIERS", "TIGES", "TIMER", "TIRAGE", "TIRER", "TISSU", "TOAST", "TOISE", "TOMBE", "TONNE", "TONTE", "TORDS", "TORDU", "TOTAL", "TOTEM", "TOUCHE", "TOURS", "TRACE", "TRAIT", "TRAME", "TREVE", "TRIBU", "TRIER", "TRIOS", "TROIS", "TROPE", "TROUS", "TRUC", "TUBES", "TUILE", "TUNER", "TYPES",
		// U-words
		"UNITE", "ULTRA", "UNION", "USAGE", "USINE", "UTILE",
		// V-words
		"VENIR", "VERRE", "VILLE", "VIVRE", "VOICI", "VOILA", "VOTER", "VACHE", "VAGUE", "VAINS", "VALSE", "VALVE", "VANNE", "VARVE", "VASTE", "VEAUX", "VEDETTE", "VEILLE", "VEINE", "VELOS", "VENDS", "VENIR", "VENTS", "VENUE", "VENUS", "VERBE", "VERGE", "VERNI", "VERRA", "VERSA", "VERSE", "VERTE", "VERTS", "VERTU", "VESTE", "VIANDE", "VIDEO", "VIDER", "VIENT", "VIEUX", "VIGIE", "VIGNE", "VILLA", "VINGT", "VIOLE", "VIRAGE", "VIRUS", "VISER", "VISITE", "VITAL", "VITRE", "VIVACE", "VIVENT", "VIVES", "VOEUX", "VOILE", "VOIRE", "VOLET", "VOMIR", "VOTRE", "VOUER", "VULVE",
		// Z-words
		"ZEROS", "ZESTE",
	} {
		lexicon.AddWord(w)
	}

	// Common 6-letter words
	for _, w := range []string{"ACTEUR", "ANCIEN", "ANNEAU", "AUTOUR", "BATEAU", "BESOIN", "BUREAU", "CACHER", "CHEVAL", "COMPTE", "DONNER", "ENTRER", "ETROIT", "FIGURE", "GAMMES", "GAUCHE", "HABITE", "JARDIN", "MADAME", "MAISON", "MARCHE", "MINUTE", "MONTRE", "NATURE", "NIVEAU", "NOMBRE", "OFFRIR", "PARLER", "PENSEE", "PIERRE", "PORTER", "POSTAL", "PROPRE", "REGARD", "RETOUR", "RIVAGE", "SECOND", "SERVIR", "SOCIAL", "SORTER", "SUCCÈS", "SUIVRE", "TALENT", "VOYAGE"} {
		lexicon.AddWord(w)
	}

	// Common 7-letter words
	for _, w := range []string{"ABRITER", "ACTRICE", "ADRESSE", "AFFAIRE", "AIMABLE", "ANIMALE", "ATTENTE", "BALANCE", "BATISSE", "CAPITAL", "CAPABLE", "CHAMBRE", "CHANGER", "CHARGER", "CHERCHE", "COLLEGE", "COMPLET", "CONCERT", "COSTUME", "CULTURE", "DEMANDE", "DERNIER", "DESSERT", "EMOTION", "ENERGIE", "ENQUETE", "ENTIERE", "ENVIRON", "ETOILES", "EXEMPLE", "FAMILLE", "FEMININ", "FINANCE", "FORTUNE", "GENERAL", "GRATUIT", "GRANDIR", "HAUTEUR", "HISTOIRE", "HOPITAL", "HUMAINE", "INCENDIE", "INSTANT", "JUSTICE", "LANGAGE", "LECTURE", "LUMIERE", "MACHINE", "MAGASIN", "MATIERE", "MEMOIRE", "MESSAGE", "MODERNE", "MONTAGE", "MUSIQUE", "MYSTERE", "NATUREL", "NOUVEAU", "NOURRIR", "OPINION", "ORIGINE", "PARFAIT", "PASSAGE", "PASSION", "PENDANT", "PENSEUR", "PLANETE", "PLATEAU", "POISSON", "PRESENT", "PROBLEME", "PRODUIRE", "PROFOND", "QUALITE", "RAPIDER", "REALITE", "RECETTE", "RENTRER", "RESERVE", "RESPECT", "REUNION", "SAISONS", "SCIENCE", "SERVICE", "SILENCE", "SOURIRE", "STATION", "SURFACE", "TABLEAU", "THEORIE", "TRAVAIL", "UNIVERS", "VICTOIRE", "VILLAGE", "VITRAIL", "VOYAGER"} {
		lexicon.AddWord(w)
	}

	return lexicon
}

File: python_generator/generator/lexicon.py
"""Lexicon for crossword word lookup and pattern matching."""

import re
from abc import ABC, abstractmethod
from pathlib import Path

from language.french import normalize_fr


class Lexicon(ABC):
    """Abstract base class for word lexicons."""

    @abstractmethod
    def words(self) -> list[str]:
        """Return all words in the lexicon."""
        pass

    @abstractmethod
    def match(self, pattern: str) -> list[str]:
        """Return words matching the pattern (. for any letter)."""
        pass

    @abstractmethod
    def contains(self, word: str) -> bool:
        """Check if word exists in lexicon."""
        pass

    @abstractmethod
    def add_words(self, words: list[str]) -> None:
        """Add words to the lexicon."""
        pass

    def words_by_length(self, length: int) -> list[str]:
        """Return all words of a specific length."""
        return [w for w in self.words() if len(w) == length]


class MemoryLexicon(Lexicon):
    """In-memory lexicon with pattern matching support."""

    def __init__(self, words: list[str] | None = None):
        self._words: set[str] = set()
        self._by_length: dict[int, set[str]] = {}
        self._pattern_cache: dict[str, list[str]] = {}

        if words:
            self.add_words(words)

    def words(self) -> list[str]:
        return list(self._words)

    def add_words(self, words: list[str]) -> None:
        """Add normalized words to the lexicon."""
        for word in words:
            normalized = normalize_fr(word)
            if normalized and len(normalized) >= 2:
                self._words.add(normalized)
                length = len(normalized)
                if length not in self._by_length:
                    self._by_length[length] = set()
                self._by_length[length].add(normalized)
        # Invalidate cache
        self._pattern_cache.clear()

    def match(self, pattern: str) -> list[str]:
        """Return words matching the pattern.

        Pattern uses . for any letter, e.g., "A.B.E" matches "AMBLE", "ABODE".
        """
        if pattern in self._pattern_cache:
            return self._pattern_cache[pattern]

        length = len(pattern)
        candidates = self._by_length.get(length, set())

        if not candidates:
            return []

        # Convert pattern to regex
        regex_pattern = "^" + pattern.replace(".", ".") + "$"
        regex = re.compile(regex_pattern)

        matches = [w for w in candidates if regex.match(w)]
        self._pattern_cache[pattern] = matches
        return matches

    def contains(self, word: str) -> bool:
        normalized = normalize_fr(word)
        return normalized in self._words

    def words_by_length(self, length: int) -> list[str]:
        return list(self._by_length.get(length, set()))


class HybridLexicon(Lexicon):
    """Lexicon that combines LLM-generated words with static fallback.

    Primary source: LLM-generated theme-aware words
    Fallback: Static word list for gap filling
    """

    def __init__(self, fallback_words: list[str] | None = None):
        self._primary = MemoryLexicon()  # LLM-generated words
        self._fallback = MemoryLexicon(fallback_words or [])

    def set_primary_words(self, words: list[str]) -> None:
        """Set the primary (LLM-generated) word list."""
        self._primary = MemoryLexicon(words)

    def words(self) -> list[str]:
        """Return all words (primary + fallback, deduplicated)."""
        all_words = set(self._primary.words())
        all_words.update(self._fallback.words())
        return list(all_words)

    def primary_words(self) -> list[str]:
        """Return only LLM-generated words."""
        return self._primary.words()

    def fallback_words(self) -> list[str]:
        """Return only static fallback words."""
        return self._fallback.words()

    def match(self, pattern: str) -> list[str]:
        """Return words matching pattern, preferring primary words."""
        primary_matches = self._primary.match(pattern)
        fallback_matches = self._fallback.match(pattern)

        # Deduplicate while preserving primary preference
        seen = set(primary_matches)
        result = primary_matches.copy()
        for word in fallback_matches:
            if word not in seen:
                result.append(word)
                seen.add(word)

        return result

    def contains(self, word: str) -> bool:
        return self._primary.contains(word) or self._fallback.contains(word)

    def add_words(self, words: list[str]) -> None:
        """Add words to the primary lexicon."""
        self._primary.add_words(words)

    def add_fallback_words(self, words: list[str]) -> None:
        """Add words to the fallback lexicon."""
        self._fallback.add_words(words)


def load_fallback_words(path: Path | str | None = None) -> list[str]:
    """Load fallback words from a text file (one word per line).

    If no path provided, loads from default data/fallback_words.txt.
    """
    if path is None:
        # Default to data/fallback_words.txt relative to this file
        path = Path(__file__).parent.parent / "data" / "fallback_words.txt"

    path = Path(path)
    if not path.exists():
        return DEFAULT_FRENCH_FALLBACK

    words = []
    with open(path, encoding="utf-8") as f:
        for line in f:
            word = line.strip()
            if word and not word.startswith("#"):
                words.append(word)
    return words if words else DEFAULT_FRENCH_FALLBACK


# Default French fallback words (common short words for gap filling)
DEFAULT_FRENCH_FALLBACK = [
    # 2 letters
    "AU", "CE", "DE", "DU", "EN", "ES", "ET", "EU", "IL", "JE", "LA", "LE",
    "LU", "MA", "ME", "MI", "MU", "NE", "NI", "NU", "ON", "OR", "OU", "PA",
    "PU", "SA", "SE", "SI", "SU", "TA", "TE", "TU", "UN", "VA", "VU",
    # 3 letters
    "AGE", "AIR", "AME", "AMI", "ANE", "ANS", "ART", "BAL", "BAS", "BEC",
    "BLE", "BOL", "BON", "BUS", "CAR", "CAS", "CLE", "COL", "CRI", "EAU",
    "ELU", "ERE", "ETE", "FEU", "FIL", "FIN", "FOI", "GEL", "ILE", "JEU",
    "LAC", "LIT", "LOI", "MAI", "MAL", "MER", "MIS", "MOI", "MOT", "MUR",
    "NEZ", "NOM", "OIE", "OSE", "PAS", "PEU", "PIE", "POT", "PRE", "RAT",
    "RIZ", "ROI", "RUE", "SAC", "SEC", "SOL", "SOI", "SUR", "THE", "TOI",
    "TON", "VIE", "VIN", "VOL",
    # 4 letters
    "AIDE", "AILE", "AMER", "AMIE", "ANGE", "ARME", "AUTO", "AVIS", "BAIN",
    "BANC", "BEAU", "BIEN", "BLEU", "BOIS", "BOND", "BORD", "BRAS", "CAFE",
    "CAMP", "CAPE", "CAVE", "CHEF", "CHER", "CIEL", "CIRE", "CLEF", "COIN",
    "COTE", "COUP", "COUR", "DENT", "DEUX", "DIEU", "DOUX", "DRAP", "DROIT",
    "ELAN", "ELLE", "EPEE", "FACE", "FAIT", "FETE", "FIER", "FILS", "FLOT",
    "FOIS", "FOND", "FOUR", "FUIT", "GARE", "GOUT", "GRIS", "HAUT", "HERBE",
    "HIER", "HIVER", "IDEE", "IRIS", "IVRE", "JAMBE", "JEAN", "JOLI", "JOUR",
    "JUIN", "JUPE", "JURY", "LAIT", "LAVE", "LIEN", "LIEU", "LION", "LIRE",
    "LONG", "LOUP", "LUXE", "MAIN", "MAIS", "MARC", "MARS", "MIDI", "MISE",
    "MODE", "MOIS", "MORT", "MUSE", "NAIN", "NERF", "NEUF", "NOCE", "NOIR",
    "NOTE", "NUIT", "ONDE", "OPUS", "OSER", "OURS", "PAGE", "PAIX", "PAPE",
    "PARE", "PART", "PAYS", "PEAU", "PERE", "PEUR", "PIED", "PILE", "PIPE",
    "PLAN", "PLUS", "POIL", "PONT", "PORT", "POUR", "PRIX", "PUCE", "PUIS",
    "REEL", "REIN", "RIEN", "RIRE", "RIVE", "ROBE", "ROCK", "ROSE", "ROUE",
    "SAGE", "SANG", "SANS", "SEIN", "SOIR", "SORT", "SOUS", "SUIS", "TARD",
    "TAUX", "TETE", "TOUR", "TOUS", "TRES", "TYPE", "VASE", "VENT", "VERS",
    "VIDE", "VITE", "VOIE", "VOIR", "VOUS", "VRAI", "YEUX", "ZERO", "ZONE",
    # 5 letters
    "ACIER", "ADIEU", "AGILE", "AIMER", "AMOUR", "ANCRE", "ANNEE", "ARBRE",
    "ASTRE", "AUTRE", "AVANT", "AVION", "AVOIR", "BAGUE", "BANQUE", "BELLE",
    "BLANC", "BLOND", "BOIRE", "BOITE", "BOMBE", "BONNE", "BRUIT", "CABLE",
    "CADRE", "CALME", "CAMEL", "CARTE", "CAUSE", "CHAIR", "CHAMP", "CHANT",
    "CHOSE", "CLAIR", "CLASSE", "COEUR", "COMME", "CONTE", "CORPS", "COURT",
    "CREER", "CRISE", "CROIX", "CYCLE", "DAME", "DANSE", "DEBUT", "DELTA",
    "DROIT", "ECOLE", "ECRAN", "EFFET", "ELEVE", "EMAIL", "ENFIN", "ENVIE",
    "EPAIS", "ETAGE", "ETUDE", "FAIRE", "FEMME", "FILLE", "FINAL", "FLEUR",
    "FORCE", "FORME", "FORUM", "FRAIS", "FRANC", "FRUIT", "FUSEE", "GARCE",
    "GARDE", "GELER", "GENRE", "GLACE", "GLOBE", "GRACE", "GRAIN", "GRAND",
    "GRAVE", "GROUPE", "GUIDE", "HOMME", "HOTEL", "HUILE", "IMAGE", "INDEX",
    "ISOLE", "JAUNE", "JEUNE", "JOUER", "LAMPE", "LARGE", "LAVER", "LECON",
    "LEGER", "LEVER", "LIBRE", "LIGNE", "LISTE", "LIVRE", "LOCAL", "LOUER",
    "LOURD", "MAGIE", "MAMAN", "MARCHE", "MATCH", "MEDIA", "MELON", "MERCI",
    "METRO", "MIEUX", "MILIEU", "MONDE", "MONTE", "MOTIF", "MOYEN", "MUSEE",
    "NEIGE", "NOBLE", "NOIRE", "OBJET", "OCEAN", "OFFRE", "OMBRE", "ONCLE",
    "OPERA", "ORDRE", "ORAGE", "PAIRE", "PANNE", "PARC", "PARIS", "PARMI",
    "PARTI", "PAUSE", "PAYER", "PEINE", "PENSE", "PERLE", "PETIT", "PHOTO",
    "PIECE", "PISTE", "PLACE", "PLAGE", "PLEIN", "POEME", "POIDS", "POINT",
    "POMME", "PORTE", "POSTE", "POUCE", "PRIME", "PRISE", "PROCHE", "PROIE",
    "QUART", "QUEUE", "RADIO", "RAISON", "RANGE", "REVUE", "RICHE", "RIVAL",
    "ROMAN", "ROUGE", "ROUTE", "SAINT", "SALON", "SCENE", "SEIZE", "SELON",
    "SENAT", "SERIE", "SIEGE", "SIGNE", "SOLDE", "SOMME", "SOUCI", "SPORT",
    "STAGE", "STYLE", "SUCRE", "SUITE", "SUPER", "TABLE", "TACHE", "TANTE",
    "TENIR", "TERRE", "TEXTE", "THEME", "TITRE", "TOMBE", "TOTAL", "TRACE",
    "TRAIN", "TRAIT", "TROIS", "TRUC", "UNITE", "USAGE", "USINE", "VAGUE",
    "VEINE", "VENIR", "VERRE", "VESTE", "VIEIL", "VIEUX", "VILLE", "VIVRE",
    "VOICI", "VOILA", "VOTRE", "ZEBRA",
]
